src/structure/model.jl

```diff
@@ -0,0 +1,1104 @@
+export StructuralElement,
+    AbstractAtom,
+    Atom,
+    DisorderedAtom,
+    AbstractResidue,
+    Residue,
+    DisorderedResidue,
+    Chain,
+    Model,
+    ProteinStructure,
+    StructuralElementOrList,
+    ishetatom,
+    serial,
+    atomname,
+    altlocid,
+    resname,
+    chainid,
+    resnumber,
+    inscode,
+    x,
+    y,
+    z,
+    coords,
+    occupancy,
+    tempfac,
+    element,
+    charge,
+    ishetero,
+    isdisorderedatom,
+    x!,
+    y!,
+    z!,
+    coords!,
+    defaultaltlocid,
+    defaultatom,
+    altlocids,
+    resid,
+    ishetres,
+    atomnames,
+    atoms,
+    isdisorderedres,
+    disorderedres,
+    defaultresname,
+    defaultresidue,
+    resnames,
+    resids,
+    residues,
+    modelnumber,
+    chainids,
+    chains,
+    structurename,
+    modelnumbers,
+    models,
+    defaultmodel,
+    sortchainids,
+    applyselectors,
+    applyselectors!,
+    collectresidues,
+    collectatoms,
+    countmodels,
+    countchains,
+    countresidues,
+    countatoms,
+    organise,
+    formatomlist,
+    choosedefaultaltlocid,
+    organisemodel,
+    organisestructure,
+    stdatomselector,
+    hetatomselector,
+    atomnameselector,
+    calpha_atom_names,
+    calphaselector,
+    backbone_atom_names,
+    backboneselector,
+    heavyatomselector,
+    resnameselector,
+    water_res_names,
+    waterselector,
+    stdresselector,
+    hetresselector,
+    disorderselector,
+    hydrogenselector
+
+
+"A protein structural element."
+abstract StructuralElement
+
+
+"An atom represented in a PDB file - either an `Atom` or a `DisorderedAtom`."
+abstract AbstractAtom <: StructuralElement
+
+"An atom represented by an ATOM or HETATM record from a PDB file."
+immutable Atom <: AbstractAtom
+    het_atom::Bool
+    serial::Int
+    name::ASCIIString
+    alt_loc_id::Char
+    res_name::ASCIIString
+    chain_id::Char
+    res_number::Int
+    ins_code::Char
+    coords::Vector{Float64}
+    occupancy::Float64
+    temp_fac::Float64
+    element::ASCIIString
+    charge::ASCIIString
+end
+
+"A container to hold different versions of the same atom in a PDB file."
+immutable DisorderedAtom <: AbstractAtom
+    alt_loc_ids::Dict{Char, Atom}
+    default::Char
+end
+
+
+"""
+A residue (amino acid) or other molecule represented in a PDB file - either a
+`Residue` or a `DisorderedResidue`.
+"""
+abstract AbstractResidue <: StructuralElement
+
+"A residue (amino acid) or other molecule record from a PDB file."
+immutable Residue <: AbstractResidue
+    name::ASCIIString
+    chain_id::Char
+    number::Int
+    ins_code::Char
+    het_res::Bool # Does the residue consist of hetatoms?
+    atom_list::Vector{ASCIIString}
+    atoms::Dict{ASCIIString, AbstractAtom}
+end
+
+# Constructor without atoms
+Residue(name::ASCIIString, chain_id::Char, number::Int, ins_code::Char, het_res::Bool) = Residue(
```

Comment:
I missed a problem about argument types. Generally speaking, broad (abstract) argument types are better than specific (concrete) ones. This is specified in the style guide of Julia: http://docs.julialang.org/en/latest/manual/style-guide/?highlight=argument#avoid-writing-overly-specific-types. So for example, this `ASCIIString` type can be replaced with `AbstractString` type, which accepts more various argument values like `SubString{ASCIIString}`, which you can see when you `split` strings. Likewise, `Integer` is better than `Int` in this case.